### PR TITLE
Improve phrase history mechanism

### LINKED ISSUE #660: Command history records all speech while talon is asleep
To reproduce:
- open command history (say "command history")
- sleep talon however you do this (ie "talon sleep")
- say random things

What I would expect: the random things I say don't get recorded by command history, because they are not commands.

What happens: the random things I say are recorded by command history

I'm not sure why this occurs, but I'm somewhat sure it didn't use to do this.

### LINKED ISSUE #757: Improve phrase history mechanism
instead of this:

https://github.com/knausj85/knausj_talon/blob/3e57e0165257cf07b0e21880d44a91e79cb3ef16/code/history.py#L19-L29

consider something like this:

```py
def on_phrase(j):
    global history
    words = j.get('text')
    if words:
        text = ' '.join(words)
        history.append(text)
        history = history[-setting_command_history_size.get() :]
```

### PR DESCRIPTION
**Pull request by pokey**
Fixes #757
Fixes #660

Also adds support for user-defined history phrase transformation by overriding `user.history_transform_phrase_text` action

This PR starts to move us in a direction where users have more flexibility to influence knausj behaviour without needing to have a fork, by adding an action that the user can override

Here's an [example](https://github.com/pokey/pokey_talon/blob/main/custom/code/history.py) of user-defined history processing

### CODE REVIEW

**Comment by pokey** [comment]
@knausj85 @rntz @splondike @nriley @wolfmanstout @AndreasArvidsson WDYT about this approach to moving away from users needing to fork knausj to alter its behaviour?

**Comment by rntz** [comment]
I like the general idea of making knausj more configurable without editing it, but what are the use-cases for this history rewriting hook you've added? Unless we know of something someone would like to do with it, I think it's better to keep things simple.

The one use-case I can think of is letting users decide whether to include or exclude phrases based on whether we're in sleep mode or not: but this PR hardcodes that we don't include them rather than letting the action determine that.

**Comment by AndreasArvidsson** [comment]
I definitely agree that configurability is a decide feature of knausj. I'm not a big fan of the command history and talon subtitles diverging though.

**Comment by rntz** [comment]
Oh, I just noticed that you have an example linked in the PR itself. Hm.

Well, I'd be happy to merge this, but I think it would be good to put the check for sleep mode into the body of `user.history_phrase_whatever` instead of hard-coding it.

**Comment by wolfmanstout** [comment]
On the specific question of this pattern as a way to support customization, I do think it makes sense. It's aligned with the approach in other systems I'm aware of that allow deep customization like WordPress and Emacs. I'm in favor of only breaking out functions when there's a demonstrated need otherwise it will add overhead everywhere. It'll be a judgment call where to draw the lines, of course, but this is probably a better approach for the long term health and adoption of knausj than having everyone maintain forks. 

Reflecting a bit on fork maintenance: for the most part I don't mind handling merge conflicts when there are updates as it educates me about new functionality that touches code I cared about enough to customize. The biggest issue I've bumped into is the work of creating clean pull requests, which requires me to work with a clean branch that's difficult for me to operate hands free. I suppose this pattern would let me avoid that, although there is some risk that if I only test my PR with my overrides in place that I may be accidentally depending on that behavior. I guess we could ask people to do a final "vanilla check". Just food for thought.

**Comment by splondike** [comment]
Learnt some new python syntax today: `if blah := foo(): print(blah)`.

I like the idea of making knausj more customisable without forking. As others have mentioned it'd have to be case by case though.

This specific change seems fine to me, though moving the sleep filter into the action as rntz said makes sense to me.

**Comment by rntz** [comment]
I guess the disadvantage of putting the sleep filter into the action is that then if the user overrides it they need to either do the sleep check themselves or else call actions.next. I think that's probably ok?

**Comment by pokey** [comment]
The main reason I didn't add sleep check to user-overridable function was that I wanted it to be a pure function, but maybe that's a bit silly.  I'm not opposed to putting it there, but yeah as @rntz mentions then user would need to implement it themselves

We could add a setting for the sleep-mode thing?

@AndreasArvidsson see https://github.com/talonvoice/talon/issues/348#issuecomment-1031498758

> I like the idea of making knausj more customisable without forking. As others have mentioned it'd have to be case by case though.

Yeah I think it should be driven by usage.  At some point I'm planning to do https://github.com/cursorless-dev/cursorless-vscode/discussions/547#discussioncomment-2115836

**Comment by splondike** [comment]
You could pass in whether speech was enabled as well if you wanted to keep it pure. I think purity is mainly useful for managing complexity (particularly when the type system can enforce it, i.e. Haskell). This case is pretty simple though.

I reckon if people are fancy enough to override the action then they would be fine to put in the sleep mode check as well, particularly if you pass in the state. Not fussed either way though.

**Comment by pokey** [comment]
Ok sleep filter moved into user action. Look good to everyone now?

**Review by splondike — APPROVED** [review]

File: code/history.py [hunk 1]
```diff
@@ -85,3 +79,11 @@ def history_get(number: int):
...
         num = (0 - number) - 1
         return history[num]
+
+    def history_transform_phrase_text(words: list[str]) -> Optional[str]:
+        """Transforms phrase text for presentation in history. Return `None` to omit from history"""
+
+        if not actions.speech.enabled():
+            return None
+
+        return ' '.join(words) if words else None
```
Comment (rntz) [inline code/history.py:89]: The 'if words else None' seems redundant here. I guess the idea is to return None if the list is empty? But the empty string is falsey in python, so the test on line 21 (`if text := ...`) will catch that. Either the test on line 21 should be for `None`, or this `if words else None` should be ditched.

File: code/history.py [hunk 2]
```diff
@@ -12,20 +13,13 @@
...
-
-
 def on_phrase(j):
     global history
 
-    try:
-        val = parse_phrase(getattr(j["parsed"], "_unmapped", j["phrase"]))
-    except:
-        val = parse_phrase(j["phrase"])
+    words = j.get('text')
```
[the remark this answers is not in the document]
Comment (pokey) [reply code/history.py:19]: I haven't tested it, but yes:

> the \\ dragon junk doesn't get surfaced in the string anymore anyway

**Pull request merged**